### PR TITLE
Override close method to make sure all messages get sent

### LINKED ISSUE #8: Not all messages get sent
I've filed [issue #5448](https://github.com/elastic/logstash/issues/5448) in the main Logstash repository, thinking that it might be a bug in how Logstash handles the inflight events, but the more I think about this the more it feels as if the `stomp` output is actually telling Logstash that it has finished its job as soon as `@client.send()` is called.

Since there's no ACK, I'm wondering if the problem could not be solved in this plugin simply by making sure to get an ACK before returning from the `receive` function.

Thoughts?

### PR DESCRIPTION
**Pull request by consulthys**
This PR fixes #8 by making sure that the stomp client properly disconnects from the stomp broker and gets a chance to empty his buffers. More information on why this is needed is available [here](https://github.com/iande/onstomp/blob/master/extra_doc/UserNarrative.md#what-really-goes-down-when-you-disconnect)

it is worth noting that this PR also includes changes made in PR #4 to add support for headers.

### CODE REVIEW

File: lib/logstash/outputs/stomp.rb [hunk 1]
```diff
@@ -57,11 +66,35 @@ def register
...
-  def receive(event)
-      
 
-      @logger.debug(["stomp sending event", { :host => @host, :event => event }])
-      @client.send(event.sprintf(@destination), event.to_json)
+  public
+  def do_close
+    @logger.warn("Disconnecting from stomp broker")
+    @client.disconnect if @client.connected?
+  end # def do_close
```
Comment (ph) [inline lib/logstash/outputs/stomp.rb]: `do_close` is our internal method which calls `close` on the plugin, so to make sure we dont introduce breaking changes we should rename that method to `close`

File: lib/logstash/outputs/stomp.rb [hunk 2]
```diff
@@ -57,11 +66,35 @@ def register
...
+        events.each { |event|
+          t.send(event.sprintf(@destination), event.to_json, headers)
+        }
+      end
+    else
+      events.each { |event|
+        @logger.debug(["stomp sending event", { :host => @host, :event => event, :headers => headers }])
+        @client.send(event.sprintf(@destination), event.to_json, headers)
+      }
+    end
```
Comment (ph) [inline lib/logstash/outputs/stomp.rb]: In Logstash `multi_receive` is clear to receive multiple events,  I don't think we should support a batch and a no batch mode.  In all cases the batch mode will be more efficient for Logstash and for stomp itself.

Do you need to enable something special on the server side to enable batch mode? I dont recall that.

File: lib/logstash/outputs/stomp.rb (hunk 1 above)
Comment (consulthys) [inline lib/logstash/outputs/stomp.rb]: Oh yes, of course, you're right. I've amended this PR with that change.

I suggest to also mention that `close` method in the ["how to contribute" documentation](https://www.elastic.co/guide/en/logstash/current/_how_to_write_a_logstash_output_plugin.html)

File: lib/logstash/outputs/stomp.rb (hunk 2 above)
Comment (consulthys) [inline lib/logstash/outputs/stomp.rb]: Ok, I'm removing the batch support then. Nothing needs to be enabled on the broker side. It works out of the box.

**Comment by ph** [comment]
@consulthys Thanks for quick turn around. This plugin would really benefit to have some tests written to it.  But since we are dealing with a data loss in a shutdown scenario we should merge it, too :)

Are you also using the stomp input?

**Comment by consulthys** [comment]
Yes, I'm also using the `stomp` input and I think it could suffer from the same issue, though, if Logstash shuts down and we don't consume all messages from the queue, it's less important as they can still be consumed on next restart.

**Comment by ph** [comment]
I will close/reopen it to trigger a travis build.

**Comment by consulthys** [comment]
The error in the Travis build seems to be related to a bad Java version.

**Comment by ph** [comment]
yeah, I've created #10 to fix it :)

**Pull request closed without merging**